python/lancedb/rerankers/base.py

```diff
@@ -0,0 +1,87 @@
+import importlib
+from typing import Any
+import lancedb
+from abc import ABC, abstractmethod
+import pyarrow as pa
+from ..utils.general import LOGGER
+
+
+class Reranker(ABC):
+    @abstractmethod
+    def rerank_hybrid(
+        query_builder: "lancedb.HybridQueryBuilder",
+        vector_results: pa.Table,
+        fts_results: pa.Table,
+    ):
+        """
+        Rerank function receives the individual results from the vector and FTS search results.
+        You can choose to use any of the results to generate the final results, allowing maximum flexibility. This is mandatory to implement
+
+        Parameters
+        ----------
+        query_builder : "lancedb.HybridQueryBuilder"
+            The query builder object that was used to generate the results
+        vector_results : pa.Table
+            The results from the vector search
+        fts_results : pa.Table
+            The results from the FTS search
+        """
+        pass
+
+    def rerank_vector(
+        query_builder: "lancedb.VectorQueryBuilder", vector_results: pa.Table
+    ):
+        """
+        Rerank function receives the individual results from the vector search. This isn't mandatory to implement
+
+        Parameters
+        ----------
+        query_builder : "lancedb.VectorQueryBuilder"
+            The query builder object that was used to generate the results
+        vector_results : pa.Table
+            The results from the vector search
+        """
+        raise NotImplementedError("Vector Reranking is not implemented")
+
+    def rerank_fts(query_builder: "lancedb.FTSQueryBuilder", fts_results: pa.Table):
+        """
+        Rerank function receives the individual results from the FTS search. This isn't mandatory to implement
+
+        Parameters
+        ----------
+        query_builder : "lancedb.FTSQueryBuilder"
+            The query builder object that was used to generate the results
+        fts_results : pa.Table
+            The results from the FTS search
+        """
+        raise NotImplementedError("FTS Reranking is not implemented")
+
+    def merge_results(self, vector_results: pa.Table, fts_results: pa.Table):
+        """
```

Comment:
The thought here is that we provide a simple way to merge fts and vector search..Because rerankers will rerank the table, the scores are irrelevant. So, this simply gets all the unique rows in fts+vector results. Does that make sense @prrao87 @changhiskhan

Reply:
Don't certain rerankers take in scores? Relative score fusion is a method that's going around these days. Some users may want a combination of results + scores to do some processing with them.

Reply:
oh yeah, I meant the ones that we're using it for don't take scores into account (cross encoder & cohere). So the current base merging strategy is set(fts+vector) but users can change that too. Unless there's a better alternative?

Reply:
That's fine. Just wanted to be sure what you meant by scores as irrelevant. As long as the fts and vector search results return the scores and we don't throw them away, the user can choose what to do with them.